File: packages/types/src/acp/update.ts
/**
 * `session/update` — the streaming notification and its `SessionUpdate`
 * payload variants (todo #350). Mirrors `mainframe-types/src/acp/update.rs`;
 * see that file for which variants are scoped out (`terminal_*`, `plan_*`,
 * `available_commands_update`, `config_option_update`, `session_info_update`
 * — none are in plan task 1's frame list).
 */
import { z } from 'zod';
import { ContentChunkSchema } from './content.js';
import { ToolCallContentChunkSchema, ToolCallUpdateSchema } from './tool-call.js';

/**
 * A user/agent message or agent-thought upsert (structurally identical
 * across the three). `content` is a patch field: omitted leaves the
 * accumulated content unchanged, `null` clears it, a value replaces it
 * wholesale — chunks (not this variant) are what append.
 */
export const MessageUpsertSchema = z
  .object({
    messageId: z.string(),
    content: z.array(ContentChunkSchema.shape.content).nullish(),
    _meta: z.record(z.string(), z.unknown()).nullish(),
  })
  .loose();
export type MessageUpsert = z.infer<typeof MessageUpsertSchema>;

export const StopReasonSchema = z.enum(['end_turn', 'max_tokens', 'max_turn_requests', 'refusal', 'cancelled']);
export type StopReason = z.infer<typeof StopReasonSchema>;

export const IdleStateUpdateSchema = z
  .object({
    stopReason: StopReasonSchema.optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type IdleStateUpdate = z.infer<typeof IdleStateUpdateSchema>;

/**
 * Foreground-work state transition. `running`/`requires_action` carry only
 * an optional `_meta` upstream; this vendored subset omits it there (nothing
 * in the facade grammar needs it) and keeps it only on `idle`, where
 * `stopReason` lives. The wire tag is `state`, flattened onto the same
 * object as `session/update`'s own `sessionUpdate` tag — Rust's internally
 * tagged newtype-variant encoding does the same flattening (`update.rs`), so
 * `SessionUpdateSchema` below extends each of these three shapes directly
 * rather than nesting a `state` sub-object.
 */
const StateRunningSchema = z.object({ state: z.literal('running') }).loose();
const StateIdleSchema = IdleStateUpdateSchema.extend({ state: z.literal('idle') });
const StateRequiresActionSchema = z.object({ state: z.literal('requires_action') }).loose();

export const SessionStateSchema = z.union([StateRunningSchema, StateIdleSchema, StateRequiresActionSchema]);
export type SessionState = z.infer<typeof SessionStateSchema>;

export const CostSchema = z
  .object({
    amount: z.number(),
    currency: z.string(),
  })
  .loose();
export type Cost = z.infer<typeof CostSchema>;

/** Context-window occupancy plus cumulative cost (ACP-EVALUATION.md "What to borrow" #5). */
export const UsageUpdateSchema = z
  .object({
    used: z.number().int().nonnegative(),
    size: z.number().int().nonnegative(),
    cost: CostSchema.optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type UsageUpdate = z.infer<typeof UsageUpdateSchema>;

/**
 * Tagged on `sessionUpdate`. A plain `z.union` rather than
 * `z.discriminatedUnion`: three options here (the `state_update` family)
 * share the `state_update` discriminator value and are told apart by the
 * nested `state` tag instead, which a single-level discriminated union can't
 * express.
 */
export const SessionUpdateSchema = z.union([
  ContentChunkSchema.extend({ sessionUpdate: z.literal('user_message_chunk') }),
  MessageUpsertSchema.extend({ sessionUpdate: z.literal('user_message') }),
  ContentChunkSchema.extend({ sessionUpdate: z.literal('agent_message_chunk') }),
  MessageUpsertSchema.extend({ sessionUpdate: z.literal('agent_message') }),
  ContentChunkSchema.extend({ sessionUpdate: z.literal('agent_thought_chunk') }),
  MessageUpsertSchema.extend({ sessionUpdate: z.literal('agent_thought') }),
  StateRunningSchema.extend({ sessionUpdate: z.literal('state_update') }),
  StateIdleSchema.extend({ sessionUpdate: z.literal('state_update') }),
  StateRequiresActionSchema.extend({ sessionUpdate: z.literal('state_update') }),
  ToolCallContentChunkSchema.extend({ sessionUpdate: z.literal('tool_call_content_chunk') }),
  ToolCallUpdateSchema.extend({ sessionUpdate: z.literal('tool_call_update') }),
  UsageUpdateSchema.extend({ sessionUpdate: z.literal('usage_update') }),
]);
export type SessionUpdate = z.infer<typeof SessionUpdateSchema>;

export const UpdateSessionNotificationSchema = z
  .object({
    sessionId: z.string(),
    update: SessionUpdateSchema,
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type UpdateSessionNotification = z.infer<typeof UpdateSessionNotificationSchema>;

File: packages/core-rs/crates/mainframe-types/src/acp/update.rs
//! `session/update` — the streaming notification and its `SessionUpdate`
//! payload variants. Internally tagged on `sessionUpdate`, mirroring the
//! `LeafContent`/`DaemonEvent` pattern already used in this crate
//! (`content.rs`, `events.rs`): each variant is a newtype wrapping a
//! dedicated struct, and `#[serde(rename_all = "snake_case")]` maps the Rust
//! variant name straight onto the schema's tag string (`ToolCallUpdate` →
//! `tool_call_update`).
//!
//! Scoped to the variants the chat facade's grammar (deltas, lifecycle,
//! usage) needs: message/thought chunks and upserts, tool-call
//! updates/chunks, foreground-state transitions, and usage. `terminal_*`
//! (declined, see tool_call.rs), `plan_update`, `available_commands_update`,
//! `config_option_update`, and `session_info_update` are not modeled — none
//! are in task 1's frame list.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::content::{ContentBlock, ContentChunk, MessageId};
use super::patch;
use super::session::SessionId;
use super::tool_call::{ToolCallContentChunk, ToolCallUpdate};

/// A user/agent message or agent-thought upsert (schema `AgentMessage` /
/// `AgentThought` / `UserMessage` — structurally identical). `content` is a
/// patch field: omitted leaves the accumulated content unchanged, `null`
/// clears it, and a value replaces it wholesale (chunks, not this variant,
/// are what append).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageUpsert {
    pub message_id: MessageId,
    #[serde(default, skip_serializing_if = "patch::is_absent", with = "patch")]
    pub content: Option<Option<Vec<ContentBlock>>>,
    #[serde(
        default,
        skip_serializing_if = "patch::is_absent",
        with = "patch",
        rename = "_meta"
    )]
    pub meta: Option<Option<Value>>,
}

/// Schema `StopReason` is `end_turn | max_tokens | max_turn_requests |
/// refusal | cancelled` plus an "other" catch-all reserved for `_`-prefixed
/// implementation extensions — it has **no built-in error reason**. The spec
/// requires one ("Turns end with an explicit stop reason (completed,
/// cancelled, **error**)"; edge case: "Adapter process dies mid-turn: the
/// turn ends with an error stop reason"), so `Error` is added here as that
/// extension, namespaced per the schema's own discipline.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StopReason {
    EndTurn,
    MaxTokens,
    MaxTurnRequests,
    Refusal,
    Cancelled,
    #[serde(rename = "_mainframe.dev/error")]
    Error,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IdleStateUpdate {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub stop_reason: Option<StopReason>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

/// Foreground-work state transition (schema `StateUpdate`). `Running` and
/// `RequiresAction` carry only an optional `_meta` upstream; this vendored
/// subset omits it on those two variants (nothing in the facade grammar needs
/// it there) and keeps it only on `Idle`, where `stopReason` lives.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "state", rename_all = "snake_case")]
pub enum SessionState {
    Running,
    Idle(IdleStateUpdate),
    RequiresAction,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Cost {
    pub amount: f64,
    pub currency: String,
}

/// Context-window occupancy plus cumulative cost (ACP-EVALUATION.md "What to
/// borrow" #5) — one event for both, replacing Mainframe's ad hoc
/// `chat.contextUsage`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UsageUpdate {
    pub used: u64,
    pub size: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cost: Option<Cost>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "sessionUpdate", rename_all = "snake_case")]
pub enum SessionUpdate {
    UserMessageChunk(ContentChunk),
    UserMessage(MessageUpsert),
    AgentMessageChunk(ContentChunk),
    AgentMessage(MessageUpsert),
    AgentThoughtChunk(ContentChunk),
    AgentThought(MessageUpsert),
    StateUpdate(SessionState),
    ToolCallContentChunk(ToolCallContentChunk),
    ToolCallUpdate(ToolCallUpdate),
    UsageUpdate(UsageUpdate),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateSessionNotification {
    pub session_id: SessionId,
    pub update: SessionUpdate,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}
